src/otaclient/grpc/api_v2/servicer.py

```diff
@@ -154,6 +178,55 @@ def _add_ecu_into_response(
                 result=failure_type,
             )
             response.add_ecu(ecu_response)
+        elif isinstance(response, api_types.ClientUpdateResponse):
+            ecu_response = api_types.ClientUpdateResponseEcu(
+                ecu_id=ecu_id,
+                result=failure_type,
+            )
+            response.add_ecu(ecu_response)
+        else:
+            raise ValueError(f"invalid response type: {response}")
+
+    def _create_local_request(
+        self,
+        req_ecu: (
+            api_types.UpdateRequestEcu
+            | api_types.RollbackRequestEcu
+            | api_types.ClientUpdateRequestEcu
+        ),
+        request_cls: (
+            type[UpdateRequestV2]
+            | type[RollbackRequestV2]
+            | type[ClientUpdateRequestV2]
+        ),
+    ) -> UpdateRequestV2 | RollbackRequestV2 | ClientUpdateRequestV2:
+        if (
+            isinstance(req_ecu, api_types.UpdateRequestEcu)
+            and (request_cls is UpdateRequestV2)
+        ) or (
+            isinstance(req_ecu, api_types.ClientUpdateRequestEcu)
+            and (request_cls is ClientUpdateRequestV2)
+        ):
+            # update or client update
+            new_session_id = gen_session_id(req_ecu.version)
+            local_request = request_cls(
+                version=req_ecu.version,
+                url_base=req_ecu.url,
+                cookies_json=req_ecu.cookies,
+                session_id=new_session_id,
+            )
+        elif isinstance(req_ecu, api_types.RollbackRequestEcu) and (
+            request_cls is RollbackRequestV2
+        ):
+            # rollback
+            new_session_id = gen_session_id("__rollback")
+            local_request = request_cls(session_id=new_session_id)
+        else:
+            raise ValueError(
+                "invalid req_ecu and request_cls combination: {req_ecu}, {request_cls}"
+            )
+
+        return local_request
```

Comment:
move the existing module to this function to extract `isinstance` part.

Reply:
Good on here! An explicit external/internal API converted is ideal :+1: